src/worker/builder.rs

```diff
@@ -152,15 +152,19 @@ impl<E: 'static> WorkerBuilder<E> {
     }
 
     /// Connect to a Faktory server with a non-standard stream.
-    pub async fn connect_with<S: AsyncRead + AsyncWrite + Send + Unpin>(
+    pub async fn connect_with<S>(
         mut self,
         stream: S,
         pwd: Option<String>,
-    ) -> Result<Worker<BufStream<S>, E>, Error> {
+    ) -> Result<Worker<E>, Error>
+    where
+        S: AsyncRead + AsyncWrite + Send + Sync + Unpin + 'static,
+        BufStream<S>: Reconnect,
```

Comment:
I _think_ if we have a blanket implementation of `Reconnect` for `BufStream<S> where S: Reconnect` it should be sufficient to add `Reconnect` to the bounds on `S`

Reply:
Ditto for `Client::connect_with`

Reply:
the thing with `Reconnect` is that `Box<dyn Connection> returned from `async fn reconnect()` should already be `AsyncBufRead`. So we are accepting any `S` that is `AsyncRead + AsyncWrite + Send + Sync + Unpin + 'static'` because we then buffer it making it `AsyncBufRead` which only then qualifies it as `Reconnect`

Reply:
Ah, hmm, I think we might need two traits here then:
```rust
trait Reconnect {
    fn reconnect(&mut self) -> io::Result<BoxedConnection>;
}
trait UnbufferedConnection: AsyncRead + AsyncWrite + Send + Sync + Unpin + 'static + ReconnectUnbuffered {}
impl<T> UnbufferedConnection for T where T: AsyncRead + AsyncWrite + Send + Sync + Unpin + 'static {}
trait ReconnectUnbuffered {
    fn reconnect(&mut self) -> io::Result<Box<dyn UnbufferedConnection>>;
}
```
with
```rust
impl<S> Reconnect for BufStream<S> where S: UnbufferedConnection {}
```
it's not super pretty, but it would allow folks to bring their own buffered _or_ unbuffered connection types, and use the former directly or use the latter wrapped in a `BufStream`.

Reply:
I am afraid I am not following here.

For both `Client::connect_with` and `WorkerBuilder::connect_with` the bound is now:
```
S: AsyncBufRead + AsyncWrite + Reconnect + Send + Sync + Unpin + 'static,
```

Reply:
I think maybe I've led you astray. I read over the code again this morning, and I think I understand what you've been trying to tell me :sweat_smile: I think the only thing left that's confusing to me is how we can have
```rust
impl<S> Reconnect for BufStream<S>
```
and also
```rust
impl Reconnect for BufStream<TlsStream<tokio::net::TcpStream>>
```
and
```rust
impl Reconnect for BufStream<TlsStream<proto::BoxedConnection>>
```
It feels like that first `impl` should overlap with the other two. I'm guessing it doesn't because we define `Reconnect` ourselves and we don't have
```rust
impl Reconnect for TlsStream<tokio::net::TcpStream>
```
or
```rust
impl Reconnect for TlsStream<proto::BoxedConnection>
```
That makes me wonder though — would it be sufficient to instead have
```rust
impl Reconnect for TlsStream<S>
``
(where I don't think we even need to require `AsyncBufRead` for that `S`)?

It also feels weird that we even have
```rust
impl Reconnect for BufStream<TlsStream<proto::BoxedConnection>>
```
Because that suggests we may end up double-buffering (since `BoxedConnection` is already `AsyncBufRead`).